### PR TITLE
Reviewer Rob - Live verification VNF

### PR DESCRIPTION
**Pull request by bossmc**
Rob, can you please review my live verification VNF code?

Interesting points:

 * I've moved to my own hosted version of `joker` since the one we were using doesn't install reliably.
 * Updated all gems (side effect of the previous change...)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
-Pre-Requisites
---------------
+## Usage Options
+
+The `clearwater-live-test` framework can be run in two modes:
+
+ * as a scriptable manual regression suite, useful for checking that a deployment is working correctly and used by the Project Clearwater team to validate that newly added function works end-to-end
+ * as a continuous verification VNF which can be installed alongside a Clearwater deployment to provide continuous, service-level verification of the deployment's basic functionality
+
+We recommend you use the manual testing regression suite when ever creating a new deployment to confirm that the deployment is correctly installed and configured.  We recommend that the verification suite is installed alongside production deployments to help with early detection of any service issues that arise.
```
Comment (rkd-msw) [inline README.md]: Ellis isn't production.

File: README.md [hunk 2]
```diff
@@ -60,17 +71,23 @@ For example, to run all the call barring tests (including the international numb
 
     rake test[test.cw-ngv.com] TESTS="Call Barring*" PSTN=true
 
-Framework Structure
----------------
+### Building the Continuous Verification Package
 
-Tests in the framework are essentially short Ruby programs. These programs use the Quaff library to talk over SIP to Clearwater nodes for calls, and the rest-client library to communicate with Ellis for provisioning.
+The continuous verification tool is packaged as a debian package (build from the same codebase as the manual test suite) ready to be installed alongside your live deployment.
```
Comment (rkd-msw) [inline README.md]: "Debian" and "built"?

File: README.md [hunk 3]
```diff
@@ -60,17 +71,23 @@ For example, to run all the call barring tests (including the international numb
...
+
+### Using the Continuous Verification Tool
 
-Writing A New Test
-------------------
+The continuous verification tool uses the same configuration file as the other Clearwater nodes, `/etc/clearwater/config`.  To install the verification VNF, prepare an Ubuntu 12.04 or 14.04 machine as if to install a Clearwater node in the deployment but, at the point you would install sprout/homer/etc. instead install the package you build in the previous step (`clearwater-live-verification`).  This will install and start the verification service.
+
+The verification service produces SNMP alarms to indicate the status of the deployment.  See [our public docs](https://clearwater.readthedocs.org/en/latest/SNMP_Alarms/index.html) for how to configure Clearwater to report these alarms.
+
+The verification service runs a very cut-down collection of tests, focussing on basic functionality tests, this means that, if the verification service reports that the deployment is ok, then subscribers are capable of registering and making calls.
```
Comment (rkd-msw) [inline README.md]: Full stop before "this means that", capitalise "OK".

File: README.md [hunk 4]
```diff
@@ -60,17 +71,23 @@ For example, to run all the call barring tests (including the international numb
...
+    ./build_deb.sh
 
-The test framework is very punctilious about cleaning up after itself, so there should be no issue with running the tests in any order or running the framework multiple times on one deployment (even at the same time!).
+Which will build a `*.deb` package in the current folder.  See `build-infra/cw-deb.mk` for more options to this script (e.g. automatically publish the package to a repository server).
+
+### Using the Continuous Verification Tool
 
-Writing A New Test
-------------------
+The continuous verification tool uses the same configuration file as the other Clearwater nodes, `/etc/clearwater/config`.  To install the verification VNF, prepare an Ubuntu 12.04 or 14.04 machine as if to install a Clearwater node in the deployment but, at the point you would install sprout/homer/etc. instead install the package you build in the previous step (`clearwater-live-verification`).  This will install and start the verification service.
```
Comment (rkd-msw) [inline README.md]: Should we just put it in our repository?

File: lib/daemon.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env ruby
+
+require 'daemons'
+require 'pathname'
+require 'ffi-rzmq'
+require 'syslog'
+require_relative './live-test'
+
+def source(file, vars)
+  ENV.replace(eval(`/bin/bash -c 'source #{file} && export #{vars.join " "} && ruby -e "p ENV"'`))
```
Comment (rkd-msw) [inline lib/daemon.rb]: Seriously?

File: lib/daemon.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+path = Pathname.new("/var/log/clearwater-live-verification/");
+Daemons.run_proc("clearwater-live-verification",
+                 dir_mode: :normal,
+                 dir: path) do
+  # Initially we don't know the state of the deployment
+  raise_alarm("6000.2")
+
+  loop do
+
+    # Sort out the appropriate ENV variables from clearwater config.
```
Comment (rkd-msw) [inline lib/daemon.rb]: Worth more explanation?

File: debian/postinst [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        mkdir -p $log_directory
+    fi
+    chmod 777 $log_directory
+}
+
+case "$1" in
+    configure)
+        setup_logging
+
+        REALLY_GEM_UPDATE_SYSTEM=true gem update --system
```
Comment (rkd-msw) [inline debian/postinst]: Why?

Comment (rkd-msw) [inline debian/postinst]: This means you need internet access for this step - worth flagging?

**Comment by rkd-msw** [comment]
A few comments, mostly around docs and comments. I'll remember `pushd` and `popd` for future.

**Comment by mirw** [comment]
@rkd-msw, please can you cast your eye over this again?  I still need to set up a Jenkins build job too.

**Comment by rkd-msw** [comment]
Looks good.

**Pull request merged**